screens/onboarding/PermissionsScreen.js

```diff
@@ -0,0 +1,262 @@
+import * as Analytics from 'expo-firebase-analytics';
+import * as Location from 'expo-location';
+import * as Permissions from 'expo-permissions';
+import geo2zip from 'geo2zip';
+import PropTypes from 'prop-types';
+import React from 'react';
+import { AsyncStorage, Image, View } from 'react-native';
+import {
+  ButtonLabel,
+  FilledButtonContainer,
+  OutlinedButtonContainer,
+  Subtitle,
+  Title,
+} from '../../components/BaseComponents';
+import Colors from '../../constants/Colors';
+import Window from '../../constants/Layout';
+import { initialRegion } from '../../constants/Map';
+import RecordIds from '../../constants/RecordIds';
+import { updateCustomers } from '../../lib/airtable/request';
+import { sendTextMessage } from '../../lib/authUtils';
+import { logErrorToSentry } from '../../lib/logUtils';
+import { findDistance, getStoreData } from '../../lib/mapUtils';
+import { PermissionsContainer } from '../../styled/auth';
+import { ColumnContainer } from '../../styled/shared';
+
+const deltas = {
+  latitudeDelta: 0.01,
+  longitudeDelta: 0.01,
+};
+
+export default class PermissionsScreen extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      locationErrorMsg: null,
+      region: initialRegion,
+      stores: null,
+      store: null,
+      showDefaultStore: false,
+      step: 1,
+    };
+  }
+
+  async getLocation() {
+    await this._findCurrentLocation();
+    this.navigateStoreSelect();
+  }
+
+  _findCurrentLocation = async () => {
```

Comment:
Many of the things in this screen were copied from `MapScreen` including `_findCurrentLocation`, `_populateInitialStoresProducts`, `_orderStoresByDistance`, etc. Would it make more sense to extract all of it out to a utils file or something like that? 

Also, it feels really inefficient that all of this is happening in `PermissionsScreen` and then AGAIN when the onboarding steps are done/app opens to the map, but also I'm not quite sure how I could eliminate that repetition/somehow pass the resulting ordered store list through

Reply:
> Would it make more sense to extract all of it out to a utils file or something like that?

I think it would! You may find that it's not super convenient to decouple the statefulness from these functions, which is why Hooks attempt to solve this problem by allowing you to reuse stateful/effectful logic across components 😄 As a hook, `_findCurrentLocation` may look something like:

```tsx
function useCurrentLocation() {
  const [error, setError] = useState(null);
  const [region, setRegion] = useState(null);

  useEffect(() => {
    async getCurrentLocation() {
      try {
        const { status } = await Permissions.askAsync(Permissions.LOCATION);
        if (status !== 'granted') {
          Analytics.setUserProperty('location_permissions', 'denied');
          throw new Error('Permission to access location was denied');
        } else {
          const location = await Location.getCurrentPositionAsync({
            timeout: 3000,
          });
          const region = {
            latitude: location.coords.latitude - deltas.latitudeDelta / 3.5,
            longitude: location.coords.longitude,
            ...deltas,
          };
          setRegion(region);
          Analytics.setUserProperty('location_permissions', 'granted');
        }
      } catch (err) {
        Analytics.setUserProperty('location_permissions', 'error');
        logErrorToSentry({
          screen: 'MapScreen',
          function: 'useCurrentLocation',
          error: err,
        });
        setError(err.message);
      }
    }
    getCurrentLocation();
  });

  return { error, region };
}
```

Of course, Hooks won't quite work with class components. But maybe good as food for thought 😄